File: scripts/model_offload.py
#!/usr/bin/env python3
"""Offload a mechanical text-generation prompt to a configured third-party
model provider. Prints generated content to stdout and exits 0 on success;
prints nothing to stdout and exits 1 on any failure. A one-line status
always goes to stderr. Callers fall back to generating the content
themselves on a non-zero exit.

Constraint: offload only activates when this script is invoked with
DigiSmith's own repo as the current working directory. The profile field
that enables it lives in `profiles/<name>.yml` inside that repo, and a
non-interactive script has no way to ask for DigiSmith's repo path the way
a conversational skill can. Run from anywhere else and the offload is
skipped, with the caller falling back to in-session generation.
"""
import argparse
import json
import os
import subprocess
import sys
import urllib.error
import urllib.request

CHUTES_CREDENTIALS_SCRIPT = os.path.expanduser(
    "~/.claude/skills/chutes-ai/scripts/manage_credentials.py"
)
CHUTES_MODEL = "google/gemma-4-31B-turbo-TEE"
CHUTES_URL = "https://llm.chutes.ai/v1/chat/completions"
TIMEOUT_SECONDS = 60


def is_digismith_repo():
    """True when the current working directory is DigiSmith's own repo."""
    plugin_json_path = os.path.join(os.getcwd(), ".claude-plugin", "plugin.json")
    if not os.path.isfile(plugin_json_path):
        return False
    try:
        with open(plugin_json_path, "r", encoding="utf-8") as f:
            plugin_config = json.load(f)
    except (OSError, UnicodeDecodeError, json.JSONDecodeError):
        return False
    return plugin_config.get("name") == "digismith"


def parse_field_value(line):
    """Extract a scalar YAML value: strip inline comments and wrapping quotes."""
    return line.split(":", 1)[1].split("#", 1)[0].strip().strip("\"'")


def read_profile_provider(profile_path):
    # Step 1: Read the pointer file to get profile name
    if not os.path.isfile(profile_path):
        return None

    try:
        with open(profile_path, "r", encoding="utf-8") as f:
            profile_name = f.read().strip()
    except (OSError, UnicodeDecodeError):
        return None

    if not profile_name:
        return None

    # Step 2: Locate DigiSmith's repo by checking for .claude-plugin/plugin.json
    if not is_digismith_repo():
        return None

    # Step 3: Read profiles/<name>.yml and extract model_offload_provider field
    profiles_file = os.path.join(os.getcwd(), "profiles", f"{profile_name}.yml")
    if not os.path.isfile(profiles_file):
        return None

    try:
        with open(profiles_file, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line.startswith("model_offload_provider:"):
                    return parse_field_value(line)
    except (OSError, UnicodeDecodeError):
        return None

    return None


def get_chutes_api_key():
    if not os.path.isfile(CHUTES_CREDENTIALS_SCRIPT):
        return None
    try:
        result = subprocess.run(
            [sys.executable, CHUTES_CREDENTIALS_SCRIPT, "get", "--field", "api_key"],
            capture_output=True,
            text=True,
            timeout=10,
        )
    except (subprocess.SubprocessError, OSError):
        return None
    if result.returncode != 0:
        return None
    key = result.stdout.strip()
    return key or None


def call_chutes(prompt, api_key):
    body = json.dumps(
        {
            "model": CHUTES_MODEL,
            "messages": [{"role": "user", "content": prompt}],
        }
    ).encode("utf-8")
    req = urllib.request.Request(
        CHUTES_URL,
        data=body,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=TIMEOUT_SECONDS) as resp:
        payload = json.loads(resp.read().decode("utf-8"))
    return payload["choices"][0]["message"]["content"]


def has_expected_html_shape(content):
    """Roughly validate the model returned a bare, fully-substituted HTML doc."""
    text = content.strip()
    if not text.lower().startswith("<!doctype html"):
        return False
    if not text.endswith("</html>"):
        return False
    if "{{" in text:
        return False
    if "```" in text:
        return False
    return True


def offload(prompt, profile_path):
    provider = read_profile_provider(profile_path)
    if provider != "chutes":
        if not is_digismith_repo():
            return None, "offload: skipped (not DigiSmith's own repo — offload only runs there)"
        return None, f"offload: skipped (model_offload_provider={provider!r}, only 'chutes' is implemented)"

    api_key = get_chutes_api_key()
    if not api_key:
        return None, "offload: skipped (no chutes credentials found)"

    try:
        content = call_chutes(prompt, api_key)
    except urllib.error.HTTPError as e:
        return None, f"offload: failed (HTTP {e.code})"
    except urllib.error.URLError as e:
        return None, f"offload: failed (network error: {e})"
    except (
        KeyError,
        IndexError,
        TypeError,
        AttributeError,
        UnicodeDecodeError,
        json.JSONDecodeError,
    ) as e:
        return None, f"offload: failed (malformed response: {e})"
    except OSError as e:
        return None, f"offload: failed (connection error: {e})"

    if not isinstance(content, str) or not content.strip():
        return None, "offload: failed (empty response)"

    if not has_expected_html_shape(content):
        return None, "offload: failed (malformed HTML shape)"

    return content, f"offload: success (chutes/{CHUTES_MODEL})"


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--prompt-file", required=True)
    parser.add_argument("--profile-path", default=".digismith/profile")
    args = parser.parse_args()

    try:
        with open(args.prompt_file, "r", encoding="utf-8") as f:
            prompt = f.read()
    except OSError as e:
        print(f"offload: failed (cannot read prompt file: {e})", file=sys.stderr)
        sys.exit(1)

    content, status = offload(prompt, args.profile_path)
    print(status, file=sys.stderr)

    if content is None:
        sys.exit(1)

    print(content)
    sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/model_offload.ts
// Offload only activates from DigiSmith's own repo: profiles/<name>.yml lives
// there, and a non-interactive script has no way to ask for DigiSmith's path
// the way a conversational skill can. Run from anywhere else and the caller
// falls back to in-session generation.
import { readFileSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { pathToFileURL } from "node:url";
import { parseArgs, requireArgs } from "./cli-args.ts";
import { resolveProvider } from "./providers/registry.ts";
import type { GatewayProvider } from "./providers/types.ts";

const TIMEOUT_MS = 60_000;

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function readUtf8Strict(path: string): string {
  return new TextDecoder("utf-8", { fatal: true }).decode(readFileSync(path));
}

export function isDigismithRepo(): boolean {
  const pluginJsonPath = join(process.cwd(), ".claude-plugin", "plugin.json");
  if (!isFile(pluginJsonPath)) return false;
  try {
    const config = JSON.parse(readUtf8Strict(pluginJsonPath));
    return config?.name === "digismith";
  } catch {
    return false;
  }
}

export function parseFieldValue(line: string): string {
  const colonIndex = line.indexOf(":");
  const afterColon = colonIndex === -1 ? "" : line.slice(colonIndex + 1);
  const hashIndex = afterColon.indexOf("#");
  const beforeComment = hashIndex === -1 ? afterColon : afterColon.slice(0, hashIndex);
  return beforeComment.trim().replace(/^['"]+|['"]+$/g, "");
}

export function readProfileProvider(profilePath: string): string | null {
  if (!isFile(profilePath)) return null;

  let profileName: string;
  try {
    profileName = readUtf8Strict(profilePath).trim();
  } catch {
    return null;
  }
  if (!profileName) return null;

  if (!isDigismithRepo()) return null;

  const profilesFile = join(process.cwd(), "profiles", `${profileName}.yml`);
  if (!isFile(profilesFile)) return null;

  let content: string;
  try {
    content = readUtf8Strict(profilesFile);
  } catch {
    return null;
  }

  for (const rawLine of content.split("\n")) {
    const line = rawLine.trim();
    if (line.startsWith("model_offload_provider:")) {
      return parseFieldValue(line);
    }
  }
  return null;
}

// Same plain-text parsing semantics as packages/jira-client/src/client.ts's
// parseEnvFile: split each line on the first "=", strip matching leading/
// trailing quotes, skip blank lines and "#" comments. Not imported from
// that package directly — it's a separate package — this is a local
// equivalent scoped to the one file this script reads.
function stripQuotes(value: string): string {
  if (value.length >= 2) {
    const first = value[0];
    const last = value[value.length - 1];
    if ((first === '"' && last === '"') || (first === "'" && last === "'")) {
      return value.slice(1, -1);
    }
  }
  return value;
}

function readDigismithEnvFile(): Record<string, string> {
  const path = join(homedir(), ".digismith", ".env");
  if (!isFile(path)) return {};

  let content: string;
  try {
    content = readUtf8Strict(path);
  } catch {
    return {};
  }

  const result: Record<string, string> = {};
  for (const rawLine of content.split("\n")) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;
    const eq = line.indexOf("=");
    if (eq === -1) continue;
    const key = line.slice(0, eq).trim();
    const value = stripQuotes(line.slice(eq + 1).trim());
    result[key] = value;
  }
  return result;
}

export function getCredential(provider: GatewayProvider): string | null {
  const envValue = process.env[provider.credentialEnv];
  if (envValue && envValue.trim()) return envValue.trim();

  const fileValue = readDigismithEnvFile()[provider.credentialEnv];
  return fileValue && fileValue.trim() ? fileValue.trim() : null;
}

export class HttpError extends Error {
  status: number;
  constructor(status: number) {
    super(`HTTP ${status}`);
    this.status = status;
  }
}

export async function callProvider(
  provider: GatewayProvider,
  prompt: string,
  apiKey: string,
): Promise<string> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    const response = await fetch(`${provider.baseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: provider.model("mechanical"),
        messages: [{ role: "user", content: prompt }],
      }),
      signal: controller.signal,
    });

    if (!response.ok) {
      throw new HttpError(response.status);
    }

    const payload: unknown = await response.json();
    const content = (payload as { choices?: Array<{ message?: { content?: unknown } }> })?.choices?.[0]?.message
      ?.content;
    if (typeof content !== "string") {
      throw new Error("malformed response: content is not a string");
    }
    return content;
  } finally {
    clearTimeout(timer);
  }
}

export function hasExpectedHtmlShape(content: string): boolean {
  const text = content.trim();
  if (!text.toLowerCase().startsWith("<!doctype html")) return false;
  if (!text.endsWith("</html>")) return false;
  if (text.includes("{{")) return false;
  if (text.includes("```")) return false;
  return true;
}

export type OffloadResult = [string | null, string];

export async function offload(prompt: string, profilePath: string): Promise<OffloadResult> {
  const providerName = readProfileProvider(profilePath);
  const provider = providerName ? resolveProvider(providerName) : undefined;

  if (!provider) {
    if (!isDigismithRepo()) {
      return [null, "offload: skipped (not DigiSmith's own repo — offload only runs there)"];
    }
    return [
      null,
      `offload: skipped (model_offload_provider=${JSON.stringify(providerName)}, no such registered provider)`,
    ];
  }

  const apiKey = getCredential(provider);
  if (!apiKey) {
    return [null, `offload: skipped (no ${provider.credentialEnv} credentials found)`];
  }

  let content: string;
  try {
    content = await callProvider(provider, prompt, apiKey);
  } catch (err) {
    if (err instanceof HttpError) {
      return [null, `offload: failed (HTTP ${err.status})`];
    }
    if (err instanceof DOMException && err.name === "AbortError") {
      return [null, "offload: failed (network error: timed out)"];
    }
    return [null, `offload: failed (${err instanceof Error ? err.message : String(err)})`];
  }

  if (!content.trim()) {
    return [null, "offload: failed (empty response)"];
  }

  if (!hasExpectedHtmlShape(content)) {
    return [null, "offload: failed (malformed HTML shape)"];
  }

  return [content, `offload: success (${provider.name}/${provider.model("mechanical")})`];
}

export async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  try {
    requireArgs(args, ["prompt-file"]);
  } catch (err) {
    console.error(`offload: failed (${err instanceof Error ? err.message : String(err)})`);
    process.exitCode = 1;
    return;
  }

  const profilePath = args["profile-path"] ?? ".digismith/profile";

  let prompt: string;
  try {
    prompt = readFileSync(args["prompt-file"], "utf-8");
  } catch (err) {
    console.error(
      `offload: failed (cannot read prompt file: ${err instanceof Error ? err.message : String(err)})`,
    );
    process.exitCode = 1;
    return;
  }

  const [content, status] = await offload(prompt, profilePath);
  console.error(status);

  if (content === null) {
    process.exitCode = 1;
    return;
  }

  console.log(content);
  process.exitCode = 0;
}

if (process.argv[1] && pathToFileURL(process.argv[1]).href === import.meta.url) {
  main();
}
